main/guides/zoe/contract-upgrade.md

```diff
@@ -0,0 +1,168 @@
+
+1. [Upgradable Declaration](#upgradable-declaration)
+2. [Durability](#durability)
+3. [Kinds](#kinds)
+4. [Crank](#crank)
+
+### Upgradable Declaration
+
+The new code bundle declares that it supports upgrade by exporting a `prepare` function in place of `start`.
+
+<<< @/snippets/zoe/src/02b-state-durable.js#export-prepare
+
+### Durability
+
+The 3rd argument, `baggage`, of the `prepare` function is a `MapStore`
+that provides a way to preserve state and behavior of objects
+between incarnations in a way that preserves identity of objects
+as seen from other vats:
+
+```js
+let publicFacet;
+if (!baggage.has('publicFacet')) {
+  // initial incarnation: create the object
+  publicFacet = makeCell();
+  baggage.init('publicFacet', publicFacet);
+} else {
+  // subsequent incarnation: use the object from the initial incarnation
+  publicFacet = baggage.get('publicFacet');
+}
+
+return { publicFacet };
+```
+
+The `provide` function supports a concise idiom for this find-or-create pattern:
+
+::: details import { provide } ...
+
+<<< @/snippets/zoe/src/02b-state-durable.js#import-provide-zone{2}
+
+:::
+
+<<< @/snippets/zoe/src/02b-state-durable.js#provide-cell
+
+::: details What happens if we don't use baggage?
+
+When the contract instance is restarted, it gets a fresh [heap](../js-programming/#vats-the-unit-of-synchrony), so [ordinary heap state](./contract-basics.html#state) does not survive upgrade. This implementation does not persist the effect of `E(publicFacet).set(2)` from the first incarnation:
+
+<<< @/snippets/zoe/src/02-state.js#heap-state{2}
+
+Also, it creates a fresh `publicFacet` object each time. So messages
+from clients that try to use the `publicFacet` from the first incarnation
+will fail to reach the `publicFacet` created in later incarnations.
+
+<<< @/snippets/zoe/src/02-state.js#fresh-export{2}
+
+:::
+
+### Kinds
+
+Use `zone.exoClass()` to define state and methods of kinds of durable objects such as `ValueCell`:
+
+::: details import { makeDurableZone } ...
+
+<<< @/snippets/zoe/src/02b-state-durable.js#import-provide-zone{1}
+
+:::
+
+<<< @/snippets/zoe/src/02b-state-durable.js#exo
+
+Now we have all the parts of an upgradable contract.
+
+::: details full contract listing
+
+<<< @/snippets/zoe/src/02b-state-durable.js#contract
+
+:::
+
+We can then upgrade it to have another method:
+
+```js
+const CellI = M.interface('ValueCell', {
+  ...
+  incr: M.call(M.number()).returns(),
+});
+
+  const makeCell = zone.exoClass('ValueCell', CellI, () => ({ value: 0 }), {
+    ...
+    incr(delta) {
+      this.state.value += delta;
+    },
+  });
+```
+
+::: tip Notes
+
+- Once the state is defined by the `init` function (3rd arg), properties cannot be added or removed.
+- Values of state properties must be serializable.
+- Values of state properties are hardened on assignment.
+- You can replace the value of a state property (e.g. `state.zot = [...state.zot, 'last']`), but you cannot mutate it (`state.zot.push('last')`).
```

Comment:
This didn't sound right to me. I asked in #dev-kernel, and the deal is that things in state are hardened, but that doesn't make them all immutable. One. example is mapStores You can add keys or modify values in a mapStore without reassigning the store.

Reply:
good point about stores.

(the zot stuff was taken directly from swingset docs, so it's wrong there too, I guess)

Reply:
@warner agreed with that and volunteered to extend the example as above.